tcptracer-bpf.c

```diff
@@ -864,38 +766,144 @@ int kretprobe__inet_csk_accept(struct pt_regs *ctx)
 		return 0;
 	}
 
+	struct tracked_socket t = {.active = 1};
+	t.prev_send_time_ns = bpf_ktime_get_ns();
+
+	bpf_map_update_elem(&tracked_sockets, &newsk, &t, BPF_ANY);
+
 	return assert_tcp_record(newsk, status, DIRECTION_INCOMING, STATE_INITIALIZING);
 }
 
-SEC("kprobe/tcp_sendmsg")
-int kprobe__tcp_sendmsg(struct pt_regs *ctx) {
-	struct sock *sk = (struct sock *) PT_REGS_PARM1(ctx);
-	size_t size = (size_t) PT_REGS_PARM3(ctx);
-	u64 zero = 0;
+__attribute__((always_inline))
+bool parse_http_response(char *buffer, int size, int *status_code_result) {
+	const char http_marker[4] = "HTTP";
+	if (size > 11) {
+        if (memcmp(buffer, http_marker, sizeof(http_marker)) == 0) {
+            int status_code = 100 * (buffer[9] - '0') + 10 * (buffer[10] - '0') + (buffer[11] - '0');
+            if (status_code > 99 && status_code < 1000) {
+                *status_code_result = status_code;
+				return true;
+			}
+		}
+	}
+	return false;
+}
+
+__attribute__((always_inline))
+bool parse_mysql_greeting(char *buffer, int size, u16 *protocol_version_result) {
+	if (size > 4) {
+		int packet_length = buffer[0] + buffer[1] * 0x100 + buffer[2] * 0x100;
+		int packet_number = (int)buffer[3];
+
+		bool is_greeting_packet = (size - 4) == packet_length && 0 == packet_number;
+		if (is_greeting_packet) {
+			u16 protocol_version = (u16)buffer[4];
+			*protocol_version_result = protocol_version;
+			return true;
+		}
+	}
+	return false;
+}
 
-	// TODO: Add DEBUG macro so this is only printed, if enabled
-	// bpf_debug("map: tcp_send_ipv4 kprobe\n");
+#define send_mysql_greeting(_ctx, _t, _protocol_version, _timestamp, _cpu) \
```

Comment:
My lack of knowledge of C and templating is going to show here: Why do we use a template here and a function (that is always inlined) for parsing? Wouldn't a function be easier here as well?

Reply:
No smart answer, sorry =/ but we got `defined with too many args` when writing as a function. Using a template was the easy solution.